### PR TITLE
feat(mesh): add toast

### PR DESCRIPTION
**Pull request by caesarWHLee**
# Feature

Add toast to show fail and success kind message. Support multiple toasts display after one.

- Toast scenarios: [doc](https://paper.dropbox.com/doc/READr-Mesh-Web-PRDWIP-pcFIPpEQo3Xf6ZoaQ0AJp#:uid=560603182671304272492117&h2=2.8-Toast)

### CODE REVIEW

**Comment by caesarWHLee** [comment]
@nihclil @LinHeMa @tyc67  請大家對照一下顯示 toast 時機的文件確認目前都有在對的地方顯示 toast，目前 payment 的地方沒有看到付款錯誤(下圖情境)的處理邏輯，所以沒有地方可以顯示

圖片

File: packages/mesh/context/toast.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+  useEffect(() => {
+    onCloseRef.current = onClose
+  }, [onClose])
+
+  useEffect(() => {
+    // when toast prop is true show the toast, add delay for toastCount more than one
+    if (toast) {
+      setTimeout(() => {
+        setShowToast(true)
+      }, 0.5 * SECOND)
```
Comment (LinHeMa) [inline packages/mesh/context/toast.tsx]: 要不要把0.5變成變數呢？

File: packages/mesh/context/toast.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    }
+  }, [toast])
+
+  useEffect(() => {
+    // hide the toast after 3 seconds, wait a bit longer than duration to prevent toast prop update
+    if (showToast) {
+      const timer = setTimeout(() => {
+        setShowToast(false)
+        setTimeout(onCloseRef.current, 0.3 * SECOND)
```
Comment (LinHeMa) [inline packages/mesh/context/toast.tsx]: 要不要把0.3變成變數呢？
```suggestion
        setTimeout(onCloseRef.current, 0.3 * SECOND)
```

File: packages/mesh/context/toast.tsx [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    : ''
+
+  if (!toast) return null
+
+  return (
+    <div
+      className={`fixed bottom-full left-1/2 z-modal flex h-toast -translate-x-1/2 items-center gap-1 rounded-md pl-3 pr-4 transition-transform ${
+        toast?.status === 'success' ? 'bg-primary-600' : 'bg-custom-red'
+      } ${classes}`}
+    >
```
Comment (LinHeMa) [inline packages/mesh/context/toast.tsx]: 需不需要加上a11y?
```suggestion
    <div
      className={`fixed bottom-full left-1/2 z-modal flex h-toast -translate-x-1/2 items-center gap-1 rounded-md pl-3 pr-4 transition-transform ${
        toast?.status === 'success' ? 'bg-primary-600' : 'bg-custom-red'
      } ${classes}`}
      role=”alert”
    >
```

File: packages/mesh/hooks/use-story-picker.ts
Comment (tyc67) [inline packages/mesh/hooks/use-story-picker.ts:1]: removePickResponse -> 移除精選失敗 Toast [figma](https://www.figma.com/design/wqODc69zqgbu7TZFtQAmnN/2024_READr-Mesh_Web?node-id=8802-362378&t=xDQNkkVt5RGSce5G-4)

File: packages/mesh/context/toast.tsx [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  useEffect(() => {
+    // hide the toast after 3 seconds, wait a bit longer than duration to prevent toast prop update
+    if (showToast) {
+      const timer = setTimeout(() => {
+        setShowToast(false)
+        setTimeout(onCloseRef.current, 0.3 * SECOND)
+      }, 3 * SECOND)
+
+      return () => {
+        clearTimeout(timer)
```
Comment (tyc67) [inline packages/mesh/context/toast.tsx]: 提問: 是否需要獨立clearTimeout  `setTimeout(onCloseRef.current)`, 還是實際上會一次clear兩個 ?

File: packages/mesh/app/payment/[type]/[targetId]/_component/payment-info.tsx
Comment (tyc67) [inline packages/mesh/app/payment/[type]/[targetId]/_component/payment-info.tsx:1]: 交易失敗在 `SendTransaction` [Ln66](https://github.com/readr-media/Sachiel/blob/66f342776e0218d821f69fc21cb9e55c08ab257c/packages/mesh/components/alchemy/send-transaction.tsx#L66)

File: packages/mesh/context/toast.tsx (hunk 1 above)
Comment (caesarWHLee) [reply packages/mesh/context/toast.tsx]: 感謝建議，修改在 [85c87cf](https://github.com/readr-media/Sachiel/pull/880/commits/85c87cfde1fa1092f14abca9decac463b7106011)

File: packages/mesh/context/toast.tsx (hunk 2 above)
Comment (caesarWHLee) [reply packages/mesh/context/toast.tsx]: 感謝建議，修改在 [85c87cf](https://github.com/readr-media/Sachiel/pull/880/commits/85c87cfde1fa1092f14abca9decac463b7106011)

File: packages/mesh/context/toast.tsx (hunk 3 above)
Comment (caesarWHLee) [reply packages/mesh/context/toast.tsx]: 增加在 [c7e91ad](https://github.com/readr-media/Sachiel/pull/880/commits/c7e91ad63a979cdf62b29316e757ae56e261e976)

File: packages/mesh/hooks/use-story-picker.ts
Comment (caesarWHLee) [reply packages/mesh/hooks/use-story-picker.ts:1]: 添加在 [9c6b404](https://github.com/readr-media/Sachiel/pull/880/commits/9c6b4046151abec0208f1002d662a66d84830279)

File: packages/mesh/app/payment/[type]/[targetId]/_component/payment-info.tsx
Comment (caesarWHLee) [reply packages/mesh/app/payment/[type]/[targetId]/_component/payment-info.tsx:1]: 添加在 [9c6b404](https://github.com/readr-media/Sachiel/pull/880/commits/9c6b4046151abec0208f1002d662a66d84830279)

File: packages/mesh/context/toast.tsx (hunk 4 above)
Comment (caesarWHLee) [reply packages/mesh/context/toast.tsx]: 感謝提醒，重構在 [85c87cf](https://github.com/readr-media/Sachiel/pull/880/commits/85c87cfde1fa1092f14abca9decac463b7106011)，該 component 三個 timer 都加上 clear function

**Review by LinHeMa — APPROVED** [review]
LGTM 👍

**Review by nihclil — APPROVED** [review]

**Review by tyc67 — APPROVED** [review]

**Pull request merged**